### DISCUSSION TITLE
Consider adding `clojure.tools.logging` to babashka

### DISCUSSION THREAD

**Discussion by dharrigan** [babashka/babashka#926] [Ideas]
At the moment, I'm using a library and trying to invoke it as part of babashka tasks, but it falls over since the `clojure.tools.logging` isn't available, i.e.,

```
❯ bb run docker-compose
----- Error --------------------------------------------------------------------
Type:     clojure.lang.ExceptionInfo
Message:  Could not resolve symbol: clojure.lang.LockingTransaction/isRunning
Location: clojure/tools/logging.clj:59:15
Phase:    analysis

----- Context ------------------------------------------------------------------
55:   [logger level throwable message]
56:   (if (case *force*
57:         :agent  true
58:         :direct false
59:         (and (clojure.lang.LockingTransaction/isRunning)
                  ^--- Could not resolve symbol: clojure.lang.LockingTransaction/isRunning
60:              (*tx-agent-levels* level)))
61:     (send-off *logging-agent*
62:       (fn [_#] (impl/write! logger level throwable message)))
63:     (impl/write! logger level throwable message))
64:   nil)
```

It would be super neat if babashka could support clojure.tools.logging.

**Comment by borkdude** [comment]
Can you elaborate which library is using tools.logging as a dep? I’ve also seen this with the nrepl library which I’m considering to fork to make it work with bb. Adding a logging library to bb should IMO be done because we want to have logging. But we can see if making tools.logging work from source is another option.

**Comment by dharrigan** [comment]
Hi,

An internal library, some tooling we use inside my company.

**Comment by borkdude** [comment]
@dharrigan An quick workaround for now could be to use `:bb` reader conditionals to strip out the calls to tools.logging.

**Reply by dharrigan** [reply]
Logging is useful tho, I want log information, nicely formatted to come out on the library as it's run. I don't want it to be ignored (and I don't think `println` is appropriate)

**Reply by borkdude** [reply]
@dharrigan I agree logging is useful, but logging can be implemented quite straightforwardly https://gist.github.com/borkdude/c97da85da67c7bcc5671765aef5a89ad. With clojure.tools.logging I have no idea which dependencies I'm taking on, or should take on, as it just wraps several Java implementations, I think? I've used taoensso.timbre myself in most projects for logging, but for bb imo it would be slightly overkill.

**Reply by dharrigan** [reply]
Hi,

Here are the dependencies for `c.t.l`, mostly it's just the APIs, with the implementations required during runtime:

```
https://github.com/clojure/tools.logging/blob/master/project.clj
```

**Reply by borkdude** [reply]
OK, but this means we would have to make an opinionated choice for a Java logging library if we went the tools.logging route, right? Any suggestions? Should babashka provide its own custom implementation?

**Reply by dharrigan** [reply]
Good question. I'm used to using slf4j, and there are various runtime bindings that could be used:

`http://www.slf4j.org/manual.html#swapping`

The one I normally use is the native implementation, i.e., `logback-classic`.

**Reply by borkdude** [reply]
Perhaps it wouldn't be unreasonable if babashka provides its own basic implementation?

**Reply by borkdude** [reply]
@mk mentioned pedestal.log (so did @isaksky). http://pedestal.io/api/pedestal.log/io.pedestal.log.html

**Reply by wilkerlucio** [reply]
Whatever it is, would be nice to be something with Clj and CLJS support, so we could keep portable code.

This is something I've been thinking about using in my libraries as well. Maybe it's a good opportunity for a more simplistic library.

I quite like the idea of logs as more data from https://github.com/BrunoBonacci/mulog

My issue with mulog specifically is that it only works in CLJ, so I can't use it in libraries that I target CLJS. But I believe a simpler version of it is easy to make. I've been doing one for myself that may serve as inspiration: https://github.com/wilkerlucio/log/blob/main/src/main/com/wsscode/log.cljc

**Reply by robertluo** [reply]
I would suggest [mulog](https://github.com/BrunoBonacci/mulog), in my project, we switched from timbre to mulog because the basic idea of "logging data, not string", and its clear implementation.

**Reply by dharrigan** [reply]
I quite like the idea too of `mulog` Very nice!

**Reply by borkdude** [reply]
A comment about mulog by @brunobonacci

<img width="805" alt="Screenshot 2021-07-09 at 15 44 55" src="https://user-images.githubusercontent.com/284934/125087560-debfe400-e0cc-11eb-84d8-ab3042e5f857.png">

**Comment by onetom** [comment]
To make a really educated decision, here is some related literature:
* https://lambdaisland.com/blog/2020-06-12-logging-in-clojure-making-sense-of-the-mess
* https://circleci.com/blog/how-a-simple-logging-problem-turned-into-a-bear-trap-lessons-learned/
* [μ/log and the next 100 logging systems - Bruno Bonacci](https://www.youtube.com/watch?v=P1149dWnl3k)

The [µ/log](https://github.com/BrunoBonacci/mulog) talk discusses a lot of thinking behind logging.

I think the biggest issue ppl face is that logging is often something available in the ambient environment of a process, like the random number generator or the system clock and the file system, instead of treating it as an explicit dependency of the program, like a database connection. This leads to a lot of parochial solution around how to configure, initialize, shutdown a logger instance, how to have multiple loggers. The concept of log-levels are often reified as methods or functions, as opposed to keeping it as data. The log formatting is also strongly coupled in most logging frameworks/libs. Debug-level logs are practically used for obtaining execution traces even, which makes more sense to solve in some implicit way, yet it's solved by sprinkling debug logs after almost every operation.

**Comment by onetom** [comment]
I was also recently fighting with logging systems, because the official Java 1.x AWS library (which I was using via the `cognitect.aws.api`) was spewing out a lot of useless warnings.
Here is my journey and the (partial) solution I found: https://clojurians.zulipchat.com/#narrow/stream/180378-slack-archive/topic/aws/near/244925147

the same settings didn't work when I used `com.datomic/ion-dev`...
I'm just saying, because I would love to use these software from babashka eventually too, so that would be my litmus test.

**Comment by borkdude** [comment]
I've decided to give clojure.tools.logging + timbre as the backing logging library a go.

``` 
$ ./bb -e '(ns dude) (defn foo [] (clojure.tools.logging/warn :foo)) (foo)'
2021-07-10T21:31:05.972Z MBP2019.local WARN [dude:?] - :foo

$ ./bb -e '(ns dude) (defn foo [] (taoensso.timbre/warn :foo)) (foo)'
2021-07-10T21:32:19.707Z MBP2019.local WARN [dude:1] - :foo
``` 

The increase in binary size is reasonable so far.

Please give it some testing:

Linux: https://20813-201467090-gh.circle-artifacts.com/0/release/babashka-0.4.7-SNAPSHOT-linux-amd64-static.tar.gz
macOS:  https://20808-201467090-gh.circle-artifacts.com/0/release/babashka-0.4.7-SNAPSHOT-macos-amd64.tar.gz
Windows: https://ci.appveyor.com/api/buildjobs/vsl139wt2l8lopgn/artifacts/babashka-0.4.7-SNAPSHOT-windows-amd64.zip

**Reply by borkdude** [reply]
The reasoning behind the current choice:

- clojure.tools.logging for compatibility with the ecosystem
- timbre because it's pure Clojure, so all config can be done using Clojure and no Java framework stuff

**Reply by justone** [reply]
This is fantastic, having a logging library built-in would be very useful. I tried it out today. This is my first experience using timbre and it looks like a great logging library. Most of my experience is with clojure.tools.logging backed by logback classic.

I tried scripting a bit with the binaries and I noticed that changing the levels didn't propagate to clojure.tools.logging. Here's an example:

```
#!/usr/bin/env bb-0.4.7-logging

(require '[clojure.tools.logging :as log])
(require '[taoensso.timbre :as timbre])

(defn test-fn
  []
  (log/debug "test ctl debug level")
  (log/info "test ctl info")
  (timbre/debug "test timbre debug level")
  (timbre/info "test timbre info"))

(println "before setting log level")
(test-fn)

(alter-var-root #'timbre/*config* #(assoc %1 :min-level :info))

(println "after setting log level to :info")
(test-fn)
```

The output is:
```
before setting log level
2021-07-10T23:42:52.529Z babashka DEBUG [user:?] - test ctl debug level
2021-07-10T23:42:52.531Z babashka INFO [user:?] - test ctl info
2021-07-10T23:42:52.531Z babashka DEBUG [user:10] - test timbre debug level
2021-07-10T23:42:52.531Z babashka INFO [user:11] - test timbre info
after setting log level to :info
2021-07-10T23:42:52.532Z babashka DEBUG [user:?] - test ctl debug level
2021-07-10T23:42:52.532Z babashka INFO [user:?] - test ctl info
2021-07-10T23:42:52.533Z babashka INFO [user:11] - test timbre info
```

I would have expected the debug from clojure.tools.logging not to print after changing the log level.

I also noticed that some functions weren't there, like `timbre/with-level` and `timbre/set-level!`, which would be nice for scripting.

Thanks for providing the binaries to play with, I'll keep trying things out.

**Reply by borkdude** [reply]
@justone Updated the binaries above, your feedback has been processed and your examples should work now. Also added `timbre/with-level` and `timbre/set-level!`. There might be more vars that we want to add, just let me know.

**Reply by borkdude** [reply]
Also added `infof`, etc, and `spit-appender` so you can log to files.

**Reply by justone** [reply]
@borkdude thanks for updating the binaries! I can confirm that everything works as expected. I'll report any other issues I find.

**Comment by borkdude** [comment]
Merged clojure.tools.logging + timbre to master branch.
